File: certpost/cloudflare.py
# ----------------------------------------------------------------------------------------
#   cloudflare.py
#   -------------
#
#   Cloudflare DNS API client for managing TXT records. Used for DNS-01 ACME challenges.
#   Uses a scoped API token (no IP whitelisting required).
#
#   (c) 2026 WaterJuice — Released under the Unlicense; see LICENSE.
#
#   Version History
#   ---------------
#   Mar 2026 - Created
# ----------------------------------------------------------------------------------------

# ----------------------------------------------------------------------------------------
#   Imports
# ----------------------------------------------------------------------------------------

import json
import urllib.error
import urllib.request
from typing import Any

# ----------------------------------------------------------------------------------------
#   Types
# ----------------------------------------------------------------------------------------

type JsonDict = dict[str, Any]

# ----------------------------------------------------------------------------------------
#   Constants
# ----------------------------------------------------------------------------------------

_API_BASE = "https://api.cloudflare.com/client/v4"

# ----------------------------------------------------------------------------------------
#   Cloudflare client
# ----------------------------------------------------------------------------------------


# ----------------------------------------------------------------------------------------
class CloudflareClient:
    """Client for the Cloudflare DNS API."""

    # ------------------------------------------------------------------------------------
    #   Construction
    # ------------------------------------------------------------------------------------

    def __init__(self, api_token: str, zone_id: str) -> None:
        """Initialise the Cloudflare API client."""
        self._api_token = api_token
        self._zone_id = zone_id

    # ------------------------------------------------------------------------------------
    #   Private methods
    # ------------------------------------------------------------------------------------

    # ------------------------------------------------------------------------------------
    def _api_call(
        self,
        method: str,
        path: str,
        body: JsonDict | None = None,
    ) -> JsonDict:
        """Make an API call and return the parsed JSON response."""
        url = f"{_API_BASE}{path}"
        data = json.dumps(body).encode() if body else None

        req = urllib.request.Request(
            url,
            data=data,
            headers={
                "Authorization": f"Bearer {self._api_token}",
                "Content-Type": "application/json",
            },
            method=method,
        )

        try:
            with urllib.request.urlopen(req, timeout=30) as response:
                result: JsonDict = json.loads(response.read())
        except urllib.error.HTTPError as e:
            error_body = e.read().decode()
            raise RuntimeError(f"Cloudflare API error ({e.code}): {error_body}") from e

        if not result.get("success", False):
            errors = result.get("errors", [])
            error_msgs = [str(e.get("message", "Unknown error")) for e in errors]
            raise RuntimeError(f"Cloudflare API error: {'; '.join(error_msgs)}")

        return result

    # ------------------------------------------------------------------------------------
    #   Public methods
    # ------------------------------------------------------------------------------------

    # ------------------------------------------------------------------------------------
    def set_txt_record(self, name: str, value: str) -> str:
        """Create a TXT record. Returns the record ID."""
        # First check if a record with this name already exists
        existing = self._find_txt_records(name)
        for record in existing:
            record_id = str(record.get("id", ""))
            if record_id:
                self._api_call(
                    "DELETE",
                    f"/zones/{self._zone_id}/dns_records/{record_id}",
                )

        # Create the new record
        result = self._api_call(
            "POST",
            f"/zones/{self._zone_id}/dns_records",
            {
                "type": "TXT",
                "name": name,
                "content": value,
                "ttl": 60,
            },
        )

        record_result: JsonDict = result.get("result", {})  # pyright: ignore[reportAssignmentType]
        return str(record_result.get("id", ""))

    # ------------------------------------------------------------------------------------
    def remove_txt_record(self, name: str) -> None:
        """Remove all TXT records matching the given name."""
        records = self._find_txt_records(name)
        for record in records:
            record_id = str(record.get("id", ""))
            if record_id:
                self._api_call(
                    "DELETE",
                    f"/zones/{self._zone_id}/dns_records/{record_id}",
                )

    # ------------------------------------------------------------------------------------
    def _find_txt_records(self, name: str) -> list[JsonDict]:
        """Find TXT records matching the given name."""
        result = self._api_call(
            "GET",
            f"/zones/{self._zone_id}/dns_records?type=TXT&name={name}",
        )
        records: list[JsonDict] = result.get("result", [])  # pyright: ignore[reportAssignmentType]
        return records

File: internal/dns/cloudflare.go
// ---------------------------------------------------------------------------------------
//
//	cloudflare.go
//	-------------
//
//	Cloudflare DNS API client for managing DNS records. Handles TXT records for
//	DNS-01 ACME challenges and A/CNAME records for subdomain pointing.
//
//	(c) 2026 WaterJuice — Released under the Unlicense; see LICENSE.
//
//	Version History
//	---------------
//	Apr 2026 - Created (Go rewrite)
//
// ---------------------------------------------------------------------------------------
package dns

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

const cloudflareAPIBase = "https://api.cloudflare.com/client/v4"

// CloudflareClient implements Provider for the Cloudflare DNS API.
type CloudflareClient struct {
	apiToken string
	zoneID   string
	client   *http.Client
}

// NewCloudflareClient creates a new Cloudflare DNS client.
func NewCloudflareClient(apiToken, zoneID string) *CloudflareClient {
	return &CloudflareClient{
		apiToken: apiToken,
		zoneID:   zoneID,
		client:   &http.Client{Timeout: 30 * time.Second},
	}
}

func (c *CloudflareClient) apiCall(method, path string, body any) (map[string]any, error) {
	url := cloudflareAPIBase + path
	var bodyReader io.Reader
	if body != nil {
		data, _ := json.Marshal(body)
		bodyReader = strings.NewReader(string(data))
	}

	req, err := http.NewRequest(method, url, bodyReader)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+c.apiToken)
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	respBody, _ := io.ReadAll(resp.Body)

	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("Cloudflare API error (%d): %s", resp.StatusCode, string(respBody))
	}

	var result map[string]any
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, err
	}

	if success, _ := result["success"].(bool); !success {
		errors, _ := result["errors"].([]any)
		var msgs []string
		for _, e := range errors {
			if em, ok := e.(map[string]any); ok {
				msgs = append(msgs, fmt.Sprintf("%v", em["message"]))
			}
		}
		return nil, fmt.Errorf("Cloudflare API error: %s", strings.Join(msgs, "; "))
	}

	return result, nil
}

func (c *CloudflareClient) findRecords(name, recordType string) ([]map[string]any, error) {
	path := fmt.Sprintf("/zones/%s/dns_records?type=%s&name=%s", c.zoneID, recordType, name)
	result, err := c.apiCall("GET", path, nil)
	if err != nil {
		return nil, err
	}
	records, _ := result["result"].([]any)
	var out []map[string]any
	for _, r := range records {
		if rm, ok := r.(map[string]any); ok {
			out = append(out, rm)
		}
	}
	return out, nil
}

func (c *CloudflareClient) deleteRecords(name, recordType string) error {
	records, err := c.findRecords(name, recordType)
	if err != nil {
		return err
	}
	for _, r := range records {
		id, _ := r["id"].(string)
		if id != "" {
			_, _ = c.apiCall("DELETE", fmt.Sprintf("/zones/%s/dns_records/%s", c.zoneID, id), nil)
		}
	}
	return nil
}

func (c *CloudflareClient) setRecord(recordType string, body map[string]any) (string, error) {
	name, _ := body["name"].(string)
	_ = c.deleteRecords(name, recordType)
	body["type"] = recordType
	result, err := c.apiCall("POST", fmt.Sprintf("/zones/%s/dns_records", c.zoneID), body)
	if err != nil {
		return "", err
	}
	rec, _ := result["result"].(map[string]any)
	id, _ := rec["id"].(string)
	return id, nil
}

func (c *CloudflareClient) SetTXTRecord(name, value string) (string, error) {
	return c.setRecord("TXT", map[string]any{"name": name, "content": value, "ttl": 60})
}

func (c *CloudflareClient) RemoveTXTRecord(name string) error {
	return c.deleteRecords(name, "TXT")
}

func (c *CloudflareClient) SetARecord(name, ip string) (string, error) {
	return c.setRecord("A", map[string]any{"name": name, "content": ip, "ttl": 1, "proxied": false})
}

func (c *CloudflareClient) RemoveARecord(name string) error {
	return c.deleteRecords(name, "A")
}

func (c *CloudflareClient) SetCNAMERecord(name, target string) (string, error) {
	return c.setRecord("CNAME", map[string]any{"name": name, "content": target, "ttl": 1, "proxied": false})
}

func (c *CloudflareClient) RemoveCNAMERecord(name string) error {
	return c.deleteRecords(name, "CNAME")
}
